apps/avifenc.c

```diff
@@ -243,6 +332,80 @@ static int parseU32List(uint32_t output[8], const char * arg)
     return index;
 }
 
+static avifBool parseScaleMode(avifScalingMode * mode, const char * arg)
+{
+    char buffer[128];
+    strncpy(buffer, arg, 127);
+    buffer[127] = 0;
+
+    char * axis[2] = { NULL, NULL };
+    axis[0] = strtok(buffer, ",");
+    axis[1] = strtok(NULL, ",");
+    if (strtok(NULL, ",") != NULL) {
+        return AVIF_FALSE;
+    }
+
+    mode->horizontal.d = 1;
+
+    for (int i = 0; i < 2; ++i) {
+        char * token = strtok(axis[i], "/");
+        if (token == NULL) {
+            if (i == 0) {
+                return AVIF_FALSE;
+            } else {
+                mode->vertical = mode->horizontal;
+                break;
+            }
+        }
+        avifFraction * target = (i == 0) ? &mode->horizontal : &mode->vertical;
+        target->n = atoi(token);
+        token = strtok(NULL, "/");
+        if (token != NULL) {
+            target->d = atoi(token);
+        }
+    }
+
+    return AVIF_TRUE;
+}
+
+static void moveFileSettings(avifInputFileSettings * dst, avifInputFileSettings * src)
+{
+    memset(dst, 0, sizeof(avifInputFileSettings));
+    dst->frameIndex = src->frameIndex;
+    if (src->quality) {
+        SET_FILE_SETTING(*dst, quality, *src->quality);
+    }
+    if (src->qualityAlpha) {
+        SET_FILE_SETTING(*dst, qualityAlpha, *src->qualityAlpha);
+    }
+    if (src->minQuantizer) {
+        SET_FILE_SETTING(*dst, minQuantizer, *src->minQuantizer);
+    }
+    if (src->maxQuantizer) {
+        SET_FILE_SETTING(*dst, maxQuantizer, *src->maxQuantizer);
+    }
+    if (src->minQuantizerAlpha) {
+        SET_FILE_SETTING(*dst, minQuantizerAlpha, *src->minQuantizerAlpha);
+    }
+    if (src->maxQuantizerAlpha) {
+        SET_FILE_SETTING(*dst, maxQuantizerAlpha, *src->maxQuantizerAlpha);
+    }
+    if (src->tileColsLog2) {
+        SET_FILE_SETTING(*dst, tileColsLog2, *src->tileColsLog2);
+    }
+    if (src->tileRowsLog2) {
+        SET_FILE_SETTING(*dst, tileRowsLog2, *src->tileRowsLog2);
+    }
+    if (src->autoTiling) {
+        SET_FILE_SETTING(*dst, autoTiling, *src->autoTiling);
+    }
+    if (src->scalingMode) {
+        SET_FILE_SETTING(*dst, scalingMode, *src->scalingMode);
+    }
+    dst->codecSpecificOptions = src->codecSpecificOptions;
+    memset(src, 0, sizeof(avifInputFileSettings));
```

Comment:
I understand the rationale now but it is rather complex in my opinion. Would booleans be enough instead of pointers? Also I see that the macros make it slightly less verbose but I would prefer to avoid any macro if possible.

Reply:
> Switched to boolean based.

Thank you, this is clearer.

> I can avoid the usage of `FILE_SETTINGS_ENTRY`

Please do. Structs/fields defined by macros are not easy to interact with in IDEs.

> `SET_FILE_SETTING` should be easy to understand, and expanding it will make the code very tedious.

Replacing any

```c
SET_FILE_SETTING(*fileSettings, tileRowsLog2, 0);
```
by
```c
fileSettings->tileRowsLog2 = 0;
fileSettings->tileRowsLog2Set = AVIF_TRUE;
```
is not that tedious to me. If you still prefer `SET_FILE_SETTING`, I do not mind. Getting rid of all macros is a preference I have, but keeping just this one is fine.

> Or I can make each field a `struct { int value; avifBool set }` and `setFileSettingInt()` a function if you prefer

It would be great with direct assignment `settings.tileRowsLog2 = {0, AVIF_TRUE};` but I doubt it works in C on all compiler flags of every platform we have in the CI. Alternatively it could be `settings.tileRowsLog2 = createFileSetting(0);` but it is similar to the macro above.

Up to you to go with macro, pair struct or double assignments.

Reply:
Went with the struct. Turns out it does not look that bad.